orbit_viewer/_intervals.py

```diff
@@ -0,0 +1,74 @@
+import PySide2.QtCore as QtCore
+
+import datetime as dt
+
+from typing import List, Tuple
+
+from dataclasses import dataclass, field
+
+# TODO if interval A has been selected (in list 's') and range is changed so that A is out-of-range we keep it in
+# the list of selected intervals to have it still selected if we are coming back to this range.
+# If the interval does not exists anymore (due to filter-changes) the selection is still true
+# -> needs some housekeeping
+
+
+@dataclass
+class _IntervalsData:
+    i: List[Tuple[dt.datetime, dt.datetime]] = field(default_factory=set)
+    s: List[Tuple[dt.datetime, dt.datetime]] = field(default_factory=set)
+
+
+class Intervals(QtCore.QObject):
+    """ Model containing all intervals of all trajectories """
+
+    intervals_changed = QtCore.Signal(str)
+    trajectory_deleted = QtCore.Signal(str)
+    selection_changed = QtCore.Signal(str)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._intervals = {}
+
+    def update_intervals(self, name: str, intervals: List[Tuple[dt.datetime, dt.datetime]]):
+        self._intervals.setdefault(name, _IntervalsData())
+        self._intervals[name].i = set(intervals)
+        self.intervals_changed.emit(name)
+
+    def remove_trajectory(self, name: str):
+        if name in self._intervals:
+            del self._intervals[name]
+            self.trajectory_deleted.emit(name)
+
+    def select(self, name: str, interval: Tuple[dt.datetime, dt.datetime]):
+        if name in self._intervals and \
+                interval in self._intervals[name].i and \
+                interval not in self._intervals[name].s:
+            self._intervals[name].s.add(interval)
+            self.selection_changed.emit(name)
+
+    def deselect_all(self):
+        for name, intervals in self._intervals.items():
+            if len(intervals.s) > 0:
+                intervals.s = set()
+                self.selection_changed.emit(name)
+
+    def deselect(self, name: str, interval: Tuple[dt.datetime, dt.datetime]):
+        if name in self._intervals and interval in self._intervals[name].s:
+            self._intervals[name].s.remove(interval)
+            self.selection_changed.emit(name)
+
+    def intervals(self, name: str):
+        return self._intervals[name].i
```

Comment:
Is there any reason why some functions check that "key is in dict" and some don't?

Reply:
I entirely removed all checks for `name in _intervals` and let dict() raise the exception.